File: backend/app/routers/adsense.py
"""OAuth Google AdSense — login no browser, token só no coletor."""

from __future__ import annotations

import json
import secrets
import threading
import time
from typing import Any
from urllib.parse import urlparse

from fastapi import APIRouter, HTTPException, Request
from fastapi.responses import HTMLResponse

from app.providers.adsense import auth_url, exchange_code
from app.schemas import OkResult
from app.store import load, provider as provider_cfg, update

router = APIRouter(prefix="/api/oauth/adsense", tags=["adsense"])

_LOCK = threading.Lock()
_PENDING: dict[str, dict[str, Any]] = {}
_TTL_S = 600


def _purge() -> None:
    now = time.monotonic()
    dead = [k for k, v in _PENDING.items() if now - float(v.get("at") or 0) > _TTL_S]
    for k in dead:
        _PENDING.pop(k, None)


def _listen_port(request: Request) -> int:
    return int(request.app.state.listen_port)


def _safe_return_to(raw: str | None, port: int) -> str:
    fallback = f"http://127.0.0.1:{port}/display/config"
    if not raw:
        return fallback
    parsed = urlparse(raw.strip())
    if parsed.scheme != "http" or parsed.hostname not in ("127.0.0.1", "localhost"):
        return fallback
    path = parsed.path or "/display/config"
    if not path.startswith("/display"):
        return fallback
    host = "127.0.0.1" if parsed.hostname == "localhost" else parsed.hostname
    netloc = f"{host}:{parsed.port}" if parsed.port else host
    return f"http://{netloc}{path}"


def _js_string_literal(value: str) -> str:
    """JSON-encode `value` for safe embedding inside an inline <script> block.

    Escaping just for JS string syntax is not enough: the HTML tokenizer ends
    a <script> element on the literal byte sequence "</script", regardless of
    JS string quoting. Escaping '<', '>' and '&' to \\uXXXX prevents that
    sequence (and any other HTML-significant text) from ever appearing raw.
    """
    encoded = json.dumps(value)
    return (
        encoded.replace("<", "\\u003c")
        .replace(">", "\\u003e")
        .replace("&", "\\u0026")
    )


def _html_redirect(url: str, message: str) -> HTMLResponse:
    safe = url.replace("&", "&amp;").replace('"', "&quot;")
    body = (
        "<!doctype html><meta charset=utf-8>"
        f'<meta http-equiv="refresh" content="0;url={safe}">'
        f"<script>location.replace({_js_string_literal(url)})</script>"
        f"<p>{message}</p>"
    )
    return HTMLResponse(body)


@router.get("/start", summary="Abre o login Google AdSense")
def oauth_start(request: Request, return_to: str | None = None) -> dict[str, str]:
    cfg = load()
    p = provider_cfg(cfg, "adsense")
    client_id = str(p.get("client_id") or "").strip()
    client_secret = str(p.get("client_secret") or "").strip()
    if not client_id or not client_secret:
        raise HTTPException(400, "Cole o Client ID e o Client Secret do Google Cloud antes de entrar")
    port = _listen_port(request)
    state = secrets.token_urlsafe(24)
    with _LOCK:
        _purge()
        _PENDING[state] = {"at": time.monotonic(), "return_to": _safe_return_to(return_to, port)}
    return {"url": auth_url(client_id, port, state)}


@router.get("/callback", summary="Callback OAuth (cadastrar esta URI no Google Cloud)", include_in_schema=False)
def oauth_callback(request: Request, code: str | None = None, state: str | None = None, error: str | None = None) -> HTMLResponse:
    port = _listen_port(request)
    with _LOCK:
        _purge()
        pending = _PENDING.pop(state or "", None)
    return_to = str((pending or {}).get("return_to") or _safe_return_to(None, port))
    sep = "&" if "?" in return_to else "?"
    if error:
        return _html_redirect(f"{return_to}{sep}adsense=denied", "Login Google cancelado.")
    if not code or not state or not pending:
        return _html_redirect(f"{return_to}{sep}adsense=error", "Callback OAuth inválido ou expirado.")
    cfg = load()
    p = provider_cfg(cfg, "adsense")
    client_id = str(p.get("client_id") or "").strip()
    client_secret = str(p.get("client_secret") or "").strip()
    try:
        tokens = exchange_code(client_id, client_secret, port, code)
    except RuntimeError as exc:
        return _html_redirect(f"{return_to}{sep}adsense=error", str(exc))

    def mut(cfg_now: dict[str, Any]) -> None:
        cfg_now["providers"]["adsense"]["refresh_token"] = str(tokens.get("refresh_token") or "")

    update(mut)
    return _html_redirect(f"{return_to}{sep}adsense=ok", "AdSense conectado. Pode fechar esta aba.")


@router.post("/disconnect", response_model=OkResult, summary="Esquece o login Google (mantém Client ID)")
def oauth_disconnect() -> OkResult:
    def mut(cfg: dict[str, Any]) -> None:
        cfg["providers"]["adsense"]["refresh_token"] = ""
        cfg["providers"]["adsense"]["account_name"] = ""

    update(mut)
    return OkResult(ok=True, cleared="adsense_oauth")

File: backend/src/routers/adsense.ts
import type { FastifyInstance } from "fastify";
import { randomBytes } from "node:crypto";
import { authUrl, exchangeCode } from "../providers/adsense.js";
import { load, updateSync as update, provider as providerCfg } from "../store.js";

const pending = new Map<string, { at: number; returnTo: string }>();
const TTL_S = 600;

function purge(): void {
  const now = performance.now() / 1000;
  for (const [k, v] of pending) {
    if (now - v.at > TTL_S) pending.delete(k);
  }
}

function listenPort(app: FastifyInstance): number {
  return Number((app as unknown as { listenPort?: number }).listenPort ?? 8787);
}

function safeReturnTo(raw: string | null | undefined, port: number): string {
  const fallback = `http://127.0.0.1:${port}/display/config`;
  if (!raw) return fallback;
  try {
    const parsed = new URL(raw.trim());
    if (parsed.protocol !== "http:") return fallback;
    if (parsed.hostname !== "127.0.0.1" && parsed.hostname !== "localhost") return fallback;
    const path = parsed.pathname || "/display/config";
    if (!path.startsWith("/display")) return fallback;
    const host = parsed.hostname === "localhost" ? "127.0.0.1" : parsed.hostname;
    const netloc = parsed.port ? `${host}:${parsed.port}` : host;
    return `http://${netloc}${path}${parsed.search}${parsed.hash}`;
  } catch {
    return fallback;
  }
}

function jsStringLiteral(value: string): string {
  const encoded = JSON.stringify(value);
  return encoded.replace(/</g, "\\u003c").replace(/>/g, "\\u003e").replace(/&/g, "\\u0026");
}

function htmlRedirect(url: string, message: string): string {
  const safe = url.replace(/&/g, "&amp;").replace(/"/g, "&quot;");
  return (
    `<!doctype html><meta charset=utf-8>` +
    `<meta http-equiv="refresh" content="0;url=${safe}">` +
    `<script>location.replace(${jsStringLiteral(url)})</script>` +
    `<p>${message}</p>`
  );
}

export async function createAdsenseRoutes(app: FastifyInstance): Promise<void> {
  app.get("/api/oauth/adsense/start", async (request, reply) => {
    const cfg = load() as Record<string, unknown>;
    const p = providerCfg(cfg, "adsense") as Record<string, unknown>;
    const clientId = String(p.client_id ?? "").trim();
    const clientSecret = String(p.client_secret ?? "").trim();
    if (!clientId || !clientSecret) {
      return reply.code(400).send({ ok: false, error: "Cole o Client ID e o Client Secret do Google Cloud antes de entrar" });
    }
    const port = listenPort(app);
    const state = randomBytes(18).toString("base64url");
    const query = (request.query ?? {}) as Record<string, string>;
    const returnTo = safeReturnTo(query.return_to ?? null, port);
    purge();
    pending.set(state, { at: performance.now() / 1000, returnTo });
    return { url: authUrl(clientId, port, state) };
  });

  app.get("/api/oauth/adsense/callback", async (request, reply) => {
    const query = (request.query ?? {}) as Record<string, string | undefined>;
    const code = query.code ?? null;
    const state = query.state ?? null;
    const error = query.error ?? null;
    const port = listenPort(app);
    purge();
    const pend = pending.get(state ?? "");
    if (pend) pending.delete(state ?? "");
    const returnTo = String(pend?.returnTo ?? safeReturnTo(null, port));
    const sep = returnTo.includes("?") ? "&" : "?";
    if (error) {
      const html = htmlRedirect(`${returnTo}${sep}adsense=denied`, "Login Google cancelado.");
      return reply.type("text/html").send(html);
    }
    if (!code || !state || !pend) {
      const html = htmlRedirect(`${returnTo}${sep}adsense=error`, "Callback OAuth inválido ou expirado.");
      return reply.type("text/html").send(html);
    }
    const cfg = load() as Record<string, unknown>;
    const p = providerCfg(cfg, "adsense") as Record<string, unknown>;
    const clientId = String(p.client_id ?? "").trim();
    const clientSecret = String(p.client_secret ?? "").trim();
    let tokens: Record<string, unknown>;
    try {
      tokens = await exchangeCode(clientId, clientSecret, port, code) as Record<string, unknown>;
    } catch (exc) {
      const html = htmlRedirect(`${returnTo}${sep}adsense=error`, String(exc));
      return reply.type("text/html").send(html);
    }
    update((cfgNow: Record<string, unknown>) => {
      const providers = (cfgNow.providers ?? {}) as Record<string, unknown>;
      const ads = (providers.adsense ?? {}) as Record<string, unknown>;
      ads.refresh_token = String(tokens.refresh_token ?? "");
      providers.adsense = ads;
      cfgNow.providers = providers;
    });
    const html = htmlRedirect(`${returnTo}${sep}adsense=ok`, "AdSense conectado. Pode fechar esta aba.");
    return reply.type("text/html").send(html);
  });

  app.post("/api/oauth/adsense/disconnect", async () => {
    update((cfg: Record<string, unknown>) => {
      const providers = (cfg.providers ?? {}) as Record<string, unknown>;
      const ads = (providers.adsense ?? {}) as Record<string, unknown>;
      ads.refresh_token = "";
      ads.account_name = "";
      providers.adsense = ads;
      cfg.providers = providers;
    });
    return { ok: true, cleared: "adsense_oauth" };
  });
}
